### PR TITLE
Extract apt/dnf and python dependencies in files

### PR DESCRIPTION
**Pull request by maringuu**
The comments that describe the dependencies are just copied over in the `requirements.txt` file.

Edit: Note that I didn't extract the nginx installation becaus it is optional.

### CODE REVIEW

**Review by jstucke — CHANGES_REQUESTED** [review]
This definitely needs to run on the CI before merging

File: src/helperFunctions/install.py [hunk 1]
```diff
@@ -239,3 +207,24 @@ def load_main_config() -> configparser.ConfigParser:
...
+    logging.info(f"Running: {cmd}")
+    try:
+        cmd_ = cmd if shell else cmd.split()
+        subprocess.run(cmd_, capture_output=True, encoding='UTF-8', shell=shell, **kwargs).check_returncode()
+    except CalledProcessError as err:
+        if raise_error:
+            logging.error(f"Failed to run {err.cmd}:\n{err.stderr}")
+            raise err
+        else:
+            logging.debug(f"Failed to run {err.cmd}:\n{err.stderr}\n\nIgnoring")
```
[the remark this answers is not in the document]
Comment (jstucke) [reply src/helperFunctions/install.py:230]: Also, something like "Failed to run (ignored): ..." could be easier to interpret as "Failed to run ... \n\nIgnoring" when looking at the logs

File: src/helperFunctions/install.py [hunk 2]
```diff
@@ -239,3 +207,24 @@ def load_main_config() -> configparser.ConfigParser:
...
+    stderr on non-zero returncode.
+    All keyword arguments are passed to subprocess.run.
+    """
+    # Keep in sync with run_cmd_with_logging in helperFunctions
+    # FIXME deduplicate this function
+    # We need to have shell explicitly to determine whether to split cmd or not
+    logging.info(f"Running: {cmd}")
+    try:
+        cmd_ = cmd if shell else cmd.split()
+        subprocess.run(cmd_, capture_output=True, encoding='UTF-8', shell=shell, **kwargs).check_returncode()
```
[the remark this answers is not in the document]
Comment (maringuu) [reply src/helperFunctions/install.py]: Yes! I didn't know about `check`.

File: src/helperFunctions/install.py [hunk 3]
```diff
@@ -239,3 +208,42 @@ def load_main_config() -> configparser.ConfigParser:
...
+
+def run_cmd_with_logging(cmd: str, raise_error=True, shell=False, **kwargs):
+    """
+    Runs `cmd` with subprocess.run, logs the command it executes and logs
+    stderr on non-zero returncode.
+    All keyword arguments are passed to subprocess.run.
+    """
+    logging.info(f"Running: {cmd}")
+    try:
+        # Split with shlex to have shell commands work as intendet
```
Comment (jstucke) [inline src/helperFunctions/install.py:221]: Small typo (but I don't know if the comment is really needed -- I think it is pretty self-explanatory)
```suggestion
        # Split with shlex to have shell commands work as intended
```

File: src/helperFunctions/install.py [hunk 4]
```diff
@@ -239,3 +208,42 @@ def load_main_config() -> configparser.ConfigParser:
...
+        cmd_ = cmd if shell else shlex.split(cmd)
+        subprocess.run(cmd_, capture_output=True, encoding='UTF-8', shell=shell, check=True, **kwargs)
+    except CalledProcessError as err:
+        # pylint fails to understand this code properly
+        # pylint:disable=no-else-raise
+        if raise_error:
+            logging.error(f"Failed to run {err.cmd}:\n{err.stderr}")
+            raise err
+        else:
+            logging.debug(f"Failed to run {err.cmd} (ignoring):\n{err.stderr}\n")
```
[the remark this answers is not in the document]
Comment (maringuu) [reply src/helperFunctions/install.py:231]: I see. But I still think that the code is more readable this way. (I'll change the comment later)

File: src/install/common.py [hunk 5]
```diff
@@ -36,60 +52,14 @@ def main(distribution):  # pylint: disable=too-many-statements
...
-
     install_pip('python3')
-    pip3_install_packages('setuptools')
-
-    # install general python dependencies
-    if distribution == 'fedora':
-        dnf_install_packages('file-devel')
-        dnf_install_packages('libffi-devel')
-        dnf_install_packages('python3-tlsh')
-        dnf_install_packages('python3-ssdeep')
```
[the remark this answers is not in the document]
Comment (maringuu) [reply src/install/common.py:62]: According to @any333 the reasoning was that pip had some problems on fedora.
My guess is that the problems probably do not exists anymore.

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -44,11 +44,6 @@ repos:
         language: system
         args: [--rcfile=.pylintrc]
 
--   repo: https://github.com/jumanjihouse/pre-commit-hooks
-    rev: 1.11.0
-    hooks:
-    -   id: shellcheck
-
```
Comment (jstucke) [inline .pre-commit-config.yaml:51]: this needs to be added back before the merge :sweat_smile:

**Review by jstucke — APPROVED** [review]

File: src/helperFunctions/install.py [hunk 7]
```diff
@@ -239,3 +208,45 @@ def load_main_config() -> configparser.ConfigParser:
...
+    Runs `cmd` with subprocess.run, logs the command it executes and logs
+    stderr on non-zero returncode.
+    All keyword arguments are execpt `raise_error` passed to subprocess.run.
+
+    :param raise_error: Whether or not an error should be raised when `cmd` fails
+    """
+    logging.info(f"Running: {cmd}")
+    try:
+        cmd_ = cmd if shell else shlex.split(cmd)
+        subprocess.run(cmd_, capture_output=True, encoding='UTF-8', shell=shell, check=True, **kwargs)
```
Comment (dorpvom) [inline src/helperFunctions/install.py:224]: This still isn't py < 3.7 compatible
```
14:48:11 [2021-08-11 14:48:11][install][INFO]: Running: sudo -EH pip3 install -r ./requirements_common.txt
14:48:11 Traceback (most recent call last):
14:48:11   File "./src/install.py", line 197, in <module>
14:48:11     install()
14:48:11   File "./src/install.py", line 171, in install
14:48:11     common(distribution)
14:48:11   File "/home/jenkins/workspace/bionic/src/install/common.py", line 37, in main
14:48:11     run_cmd_with_logging("sudo -EH pip3 install -r ./requirements_common.txt")
14:48:11   File "/home/jenkins/workspace/bionic/src/helperFunctions/install.py", line 224, in run_cmd_with_logging
14:48:11     subprocess.run(cmd_, capture_output=True, encoding='UTF-8', shell=shell, check=True, **kwargs)
14:48:11   File "/usr/lib/python3.6/subprocess.py", line 423, in run
14:48:11     with Popen(*popenargs, **kwargs) as process:
14:48:11 TypeError: __init__() got an unexpected keyword argument 'capture_output'
```

**Review by dorpvom — CHANGES_REQUESTED** [review]

File: src/install/common.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jstucke) [reply src/install/common.py:62]: according to @dorpvom, the problems still exists and we still need this for fedora

[the remark this answers is not in the document]
Comment (maringuu) [reply src/install/common.py:62]: Added the `ssdep-devel` dependency for fedora so the build should now succeed.

**Review by dorpvom — APPROVED** [review]

**Pull request merged**